extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/DoubleBloomFilterAggregator.java

```diff
@@ -23,20 +23,22 @@
 import org.apache.druid.query.filter.BloomKFilter;
 import org.apache.druid.segment.BaseDoubleColumnValueSelector;
 
+import java.nio.ByteBuffer;
+
 public final class DoubleBloomFilterAggregator extends BaseBloomFilterAggregator<BaseDoubleColumnValueSelector>
```

Comment:
I think you could collapse all of these implementations into a single one using a similar technique to CardinalityAggregator. (ColumnSelectorPluses encapsulating a strategy for reading different types)

No need to do it for this PR, IMO, unless you want to. Just wanted to mention the technique.

Reply:
Heh, originally the bloom filter aggregators were done like that, [refactored in the original PR in this commit](https://github.com/apache/incubator-druid/pull/6397/commits/74feb9750eeca367728a6ee60b677994fe7e18f0), but it actually takes more classes to implement it that way (especially the way it is in this PR), and puts more indirection between the aggregators which doesn't really seem necessary. See [this comment chain original PR for reference](https://github.com/apache/incubator-druid/pull/6397#discussion_r250564034), and the issue spawned from it, #6909 for more details. I think I find it cleaner this way personally, especially after the changes of this PR.

Reply:
Oh. Well, I think the way the CardinalityAggregator does it is cleaner, since it only splits out the logic for how to read from differently-typed inputs, and then composes that logic into a single Aggregator class, instead of using inheritance and a ton of Aggregator subclasses. I just generally find inheritance based structures less easy to follow since the logic seems 'inside out' to me.

But, I guess if reasonable people could differ on this point, it's up to you how you want to structure it. I don't think this inheritance based approach is _better_ than the type-based-input-strategy approach, but it's acceptable.

Reply:
By the way, one original reason for the CardinalityAggregator-style design choice is that the logic for choosing which implementation to use could be centralized / standardized. The equivalent of this code here: https://github.com/apache/incubator-druid/pull/7496/files#diff-06cafba60d560f4a5bb1551a56b8041dR99 is this code in CardinalityAggregatorFactory:

```java
    ColumnSelectorPlus<CardinalityAggregatorColumnSelectorStrategy>[] selectorPluses =
        DimensionHandlerUtils.createColumnSelectorPluses(
            STRATEGY_FACTORY,
            fields,
            columnFactory
        );
```

Which calls out to a standard library function. One other cool thing we can do here is modify the ColumnSelectorStrategyFactory interface to look more like this:

```java
public interface ColumnStrategizer<T>
{
  T makeDimensionStrategy(DimensionSelector selector);

  T makeFloatStrategy(ColumnValueSelector selector);

  T makeDoubleStrategy(ColumnValueSelector selector);

  T makeLongStrategy(ColumnValueSelector selector);
}
```

That way, there's a structure in place to make sure every type has some sort of handling. Or if we used default methods, at least consistent exceptions being thrown when types are missing handling. It isn't like this yet, but it could be, and I think it would be nice.

Reply:
I don't feel super strongly either way, but I think it is slightly fewer classes the way it is in this PR since just the base agg and the subclasses with the selector specific handling, instead of the agg, strategy factory, and strategy for each selector. Maybe there is another abstraction somewhere in between these two approaches, or maybe a way to just supply strategies to a common strategy factory so you don't have to do that strategy factory boilerplate? I'll think about this a bit more and maybe follow up in the future.

Reply:
For logic as simple as this case, the strategies could probably all be lambdas inlined inside the strategy factory.